### PR TITLE
fix(reporting): gateway events are not reported when source is connected to no destination(or device mode destination)

### PR DESCRIPTION
**Pull request by Sidddddarth**
## Description of the change

> Moved around metric computation in processor to get proper ingestion numbers.
> 1. Sources with no destinations
> 2. Sources with Device mode destinations

Example scenarios:

1. 1 source - no dest:                                gateway = 1, destination_filter = -1(diff)
2. 1 source - device mode dest:              gateway = 1, destination_filter = -1(diff)
3. 1 source - 2 dests(1 device, 1 cloud): gateway = 1, transformation = 1, router = 1

## Notion Link

> https://www.notion.so/rudderstacks/Reporting-does-not-report-Gateway-events-when-there-is-no-destination-is-connected-to-a-source-b5a37d7df9944c84818c42475d7c55ca

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: processor/processor.go [hunk 1]
```diff
@@ -879,21 +879,59 @@ func (proc *HandleT) updateMetricMaps(countMetadataMap map[string]MetricMetadata
...
+
 		if _, ok := countMap[countKey]; !ok {
 			countMap[countKey] = 0
 		}
 		countMap[countKey] = countMap[countKey] + 1
+
 		if countMetadataMap != nil {
 			if _, ok := countMetadataMap[countKey]; !ok {
-				countMetadataMap[countKey] = MetricMetadata{sourceID: event.Metadata.SourceID, destinationID: event.Metadata.DestinationID, sourceBatchID: event.Metadata.SourceBatchID, sourceTaskID: event.Metadata.SourceTaskID, sourceTaskRunID: event.Metadata.SourceTaskRunID, sourceJobID: event.Metadata.SourceJobID, sourceJobRunID: event.Metadata.SourceJobRunID, sourceDefinitionID: event.Metadata.SourceDefinitionID, destinationDefinitionID: event.Metadata.DestinationDefinitionID, sourceCategory: event.Metadata.SourceCategory}
+				countMetadataMap[countKey] = MetricMetadata{
```
Comment (atzoum) [inline processor/processor.go:898]: 👍

File: processor/processor.go [hunk 2]
```diff
@@ -879,21 +879,59 @@ func (proc *HandleT) updateMetricMaps(countMetadataMap map[string]MetricMetadata
 		var eventType string
 		eventName = event.Metadata.EventName
 		eventType = event.Metadata.EventType
-		countKey := strings.Join([]string{event.Metadata.SourceID, event.Metadata.DestinationID, event.Metadata.SourceBatchID, eventName, eventType}, METRICKEYDELIMITER)
+
+		countKey := strings.Join([]string{
```
Comment (atzoum) [inline processor/processor.go:883]: 👍

File: processor/processor.go [hunk 3]
```diff
@@ -1654,12 +1707,18 @@ func (proc *HandleT) transformSrcDest(
 
 	//REPORTING - START
 	if proc.isReportingEnabled() {
-		//Grouping events by sourceid + destinationid + source batch id to find the count
+		//Grouping events by sourceid + destinationid + sourcebatchid + eventName + eventType batch id to find the count
```
Comment (atzoum) [inline processor/processor.go:1710]:
```suggestion
		//Grouping events by sourceid + destinationid + sourcebatchid + eventName + eventType to find the count
```

**Review by atzoum — COMMENTED** [review]
Still many questions will go through the changes together

File: processor/processor.go [hunk 4]
```diff
@@ -1168,25 +1209,42 @@ func (proc *HandleT) processJobsForDest(subJobs subJob, parsedEventList [][]type
...
 				}
 
+				commonMetadataFromSingularEvent := makeCommonMetadataFromSingularEvent(
+					singularEvent,
+					batchEvent,
+					receivedAt,
+					sourceForSingularEvent,
+				)
+
+				//REPORTING - GATEWAY metrics - START
```
Comment (atzoum) [inline processor/processor.go:1230]: The current implementation of capturing reporting metrics seems to be convoluted. The least I can say is that it is hard to follow, since as it appears many things need to happen at different stages of this method, without a clear purpose.
I will invest some time to study the current implementation in hope of forming a cleaner picture for the specific domain, otherwise I cannot be of any more use with regards to reviewing this PR.

**Comment by atzoum** [comment]
Please update the pull request's title according to our [new policy](https://www.notion.so/rudderstacks/Pull-Requests-40a4c6bd7a5e4387ba9029bab297c9e3)

**Comment by Sidddddarth** [comment]
> Please update the pull request's title according to our [new policy](https://www.notion.so/rudderstacks/Pull-Requests-40a4c6bd7a5e4387ba9029bab297c9e3)

Is there an example I can see?
Because this is how I interpreted the policy wrt the title😅

From the article:
`<type>[optional scope]: <description>`
I think this is a fix, although not a bug. Maybe it's a feature. Optional scope would be Reporting here?
The `description` portion can be shortened a bit more imo. I'll do that now.

Suggestions/corrections welcome.

**Comment by atzoum** [comment]
> > Please update the pull request's title according to our [new policy](https://www.notion.so/rudderstacks/Pull-Requests-40a4c6bd7a5e4387ba9029bab297c9e3)
> 
> Is there an example I can see? Because this is how I interpreted the policy wrt the title😅
> 
> From the article: `<type>[optional scope]: <description>` I think this is a fix, although not a bug. Maybe it's a feature. Optional scope would be Reporting here? The `description` portion can be shortened a bit more imo. I'll do that now.
> 
> Suggestions/corrections welcome.

This should be enough (I used the title from the notion ticket, but feel free to change it if there is something more appropriate):
```
fix(reporting): gateway events are not reported when there is no destination connected to a source
```
In the title you are free to use language that humans can understand (with spaces) 😊

**Comment by Sidddddarth** [comment]
> This should be enough (I used the title from the notion ticket, but feel free to change it if there is something more appropriate):
> 
> ```
> fix(reporting): gateway events are not reported when there is no destination connected to a source
> ```
> 
> In the title you are free to use language that humans can understand (with spaces) 😊

Understood now.
I'd taken these examples to heart:
- `feat.pendingEvtRedesign`
- `fix.routerCrashLoop`
- `chore.ghLabelAction`
- `refactor.jobsdbQueries`
- `exp.tzoumasa`
- `doc.readme`
- `test.multitenantPickupAlg`
emphasis on human-readable.

**Review by chandumlg — APPROVED** [review]

**Review by cisse21 — APPROVED** [review]
